pkg/profiler/cpu/cpu.go

```diff
@@ -269,6 +272,76 @@ func loadBpfProgram(logger log.Logger, debugEnabled bool, memlockRlimit uint64)
 	return nil, nil, err
 }
 
+func (p *CPU) DwarfUnwindingWithoutPolling() bool {
+	return !p.disableDWARFUnwinding && !p.dwarfUnwindingPolling
+}
+
+func (p *CPU) addUnwindTableForProcess(pid int) {
+	executable := fmt.Sprintf("/proc/%d/exe", pid)
+	hasFramePointers, err := p.framePointerCache.HasFramePointers(executable)
+	if err != nil {
+		// It might not exist as reading procfs is racy.
+		if !errors.Is(err, os.ErrNotExist) {
+			level.Error(p.logger).Log("msg", "frame pointer detection failed", "executable", executable, "err", err)
+		}
+		return
+	}
+
+	if hasFramePointers {
+		return
+	}
+
+	level.Debug(p.logger).Log("msg", "adding unwind tables", "pid", pid)
+
+	err = p.bpfMaps.addUnwindTableForProcess(pid)
+	if err != nil {
+		//nolint: gocritic
+		if errors.Is(err, ErrNeedMoreProfilingRounds) {
+			level.Debug(p.logger).Log("msg", "PersistUnwindTable called to soon", "err", err)
+		} else if errors.Is(err, os.ErrNotExist) {
+			level.Debug(p.logger).Log("msg", "failed to add unwind table due to a procfs race", "pid", pid, "err", err)
+		} else if errors.Is(err, errTooManyExecutableMappings) {
+			level.Warn(p.logger).Log("msg", "failed to add unwind table due to having too many executable mappings", "pid", pid, "err", err)
+		} else {
+			level.Error(p.logger).Log("msg", "failed to add unwind table", "pid", pid, "err", err)
+		}
+		return
+	}
+}
+
+// onDemandUnwindInfoBatcher batches PIDs sent from the BPF program when
+// frame pointers and unwind information are not present.
+//
+// Waiting for as long as `duration` is important because `PersistUnwindTable`
+// must be called to write the in-flight shard to the BPF map. This has been
+// a hot path in the CPU profiles we take in Demo when we persisted the unwind
+// tables after adding every pid.
+func onDemandUnwindInfoBatcher(eventsChannel <-chan []byte, duration time.Duration, callback func([]int)) {
+	batch := make([]int, 0)
+	timerOn := false
+	timer := &time.Timer{}
```

Comment:
I suspect this is to protect from `timer` not being truly set yet and for example on line 336 on the first iteration it would result in a `nil` panic. I would have instead expected to see just a variable declaration and a first `select` statement outside of the `for` loop.

Reply:
Exactly, yeah! Thought about doing it in two selects, but found it less clear. No strong opinions though!